Source/SwiftLintBuiltInRules/Rules/Metrics/NestingRule.swift

```diff
@@ -12,137 +13,171 @@ struct NestingRule: Rule {
+            return .visitChildren
         }
 
-        return validate(file: file, substructure: dictionary.substructure,
-                        args: ValidationArgs(
-                            typeLevel: currentTypeLevel,
-                            functionLevel: currentFunctionLevel,
-                            previousKind: structureKind,
-                            violations: violations
-            )
-        )
-    }
+        override func visitPost(_ node: EnumDeclSyntax) {
+            validationData.pop()
+        }
 
-    private func levelViolation(file: SwiftLintFile, dictionary: SourceKittenDictionary,
-                                previousKind: SwiftStructureKind?, level: Int, forFunction: Bool) -> StyleViolation? {
-        guard let offset = dictionary.offset else {
-            return nil
+        override func visit(_ node: ExtensionDeclSyntax) -> SyntaxVisitorContinueKind {
+            validate(node, inFunction: validationData.isFunction)
+            return .visitChildren
         }
 
-        let targetLevel = forFunction ? configuration.functionLevel : configuration.typeLevel
-        var violatingSeverity: ViolationSeverity?
+        override func visitPost(_ node: ExtensionDeclSyntax) {
+            validationData.pop()
+        }
 
-        if configuration.alwaysAllowOneTypeInFunctions,
-            case let .declaration(previousDeclarationKind)? = previousKind,
-            !SwiftDeclarationKind.functionKinds.contains(previousDeclarationKind) {
-            violatingSeverity = configuration.severity(with: targetLevel, for: level)
-        } else if forFunction || !configuration.alwaysAllowOneTypeInFunctions || previousKind == nil {
-            violatingSeverity = configuration.severity(with: targetLevel, for: level)
-        } else {
-            violatingSeverity = nil
+        override func visit(_ node: FunctionDeclSyntax) -> SyntaxVisitorContinueKind {
+            validate(node, triggeringToken: node.funcKeyword, inFunction: validationData.isFunction)
+            return .visitChildren
+        }
+
+        override func visitPost(_ node: FunctionDeclSyntax) {
+            validationData.pop()
+        }
+
+        override func visit(_ node: StructDeclSyntax) -> SyntaxVisitorContinueKind {
+            validate(node, triggeringToken: node.structKeyword, inFunction: validationData.isFunction)
+            return .visitChildren
+        }
+
+        override func visitPost(_ node: StructDeclSyntax) {
+            validationData.pop()
+        }
+
+        // MARK: - configuration for ignoreTypealiasesAndAssociatedtypes
+        override func visitPost(_ node: TypeAliasDeclSyntax) {
+            guard !configuration.ignoreTypealiasesAndAssociatedtypes else { return }
+            validate(node, triggeringToken: node.typealiasKeyword, inFunction: false)
+            validationData.pop()
+        }
+
+        // associatedtype is only permitted within a protocol,
```

Comment:
That's true for Swift versions < 5.10. Swift 5.10 [allows](https://github.com/apple/swift-evolution/blob/main/proposals/0404-nested-protocols.md) to nest protocols inside of other non-generic contexts. Only then, the option starts to make sense eventually.